File: backend/api/app/models/ai_alert.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, String, Text, func
from app.database import GUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class AlertSeverity(str, enum.Enum):
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"


class AiAlert(Base):
    __tablename__ = "ai_alerts"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("shipments.id"), nullable=True)
    alert_type: Mapped[str] = mapped_column(String(100), nullable=False)
    severity: Mapped[AlertSeverity] = mapped_column(Enum(AlertSeverity), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    meta_data: Mapped[str | None] = mapped_column("metadata", Text, nullable=True)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

File: packages/shared/lib/src/models/ai_alert.dart
class AiAlert {
  final String id;
  final String? shipmentId;
  final String alertType;
  final String severity;
  final String title;
  final String description;
  final Map<String, dynamic>? metadata;
  final bool isRead;
  final DateTime createdAt;

  AiAlert({
    required this.id,
    this.shipmentId,
    required this.alertType,
    required this.severity,
    required this.title,
    required this.description,
    this.metadata,
    required this.isRead,
    required this.createdAt,
  });

  factory AiAlert.fromJson(Map<String, dynamic> json) {
    return AiAlert(
      id: json['id'] as String,
      shipmentId: json['shipment_id'] as String?,
      alertType: json['alert_type'] as String,
      severity: json['severity'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      metadata: json['metadata'] as Map<String, dynamic>?,
      isRead: json['is_read'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'shipment_id': shipmentId,
      'alert_type': alertType,
      'severity': severity,
      'title': title,
      'description': description,
      'metadata': metadata,
      'is_read': isRead,
      'created_at': createdAt.toIso8601String(),
    };
  }
}
